crates/accelerate/src/sparse_pauli_op.rs

```diff
@@ -337,163 +338,463 @@ pub fn decompose_dense(
+        let col = loc.col();
+        let base = row * side + col;
+        let i_value = scratch[base] + scratch[base + side + 1];
+        let z_value = scratch[base] - scratch[base + side + 1];
+        let x_value = scratch[base + 1] + scratch[base + side];
+        let y_value = scratch[base + 1] - scratch[base + side];
+
+        let x = row ^ col;
+        let z = row;
+        let phase = (x & z).count_ones() as u8;
+        // Pushing the last Pauli onto the `loc` happens "forwards" to maintain lexicographical
+        // ordering in `out`, since this is the construction of the final object.
+        push_pauli_if_nonzero(x, z, phase, i_value, &mut out);
+        push_pauli_if_nonzero(x | 1, z, phase, x_value, &mut out);
+        push_pauli_if_nonzero(x | 1, z | 1, phase + 1, y_value, &mut out);
+        push_pauli_if_nonzero(x, z | 1, phase, z_value, &mut out);
+    }
+    // If we _wildly_ overallocated, then shrink back to a sensible size to avoid tying up too much
+    // memory as we return to Python space.
+    if out.z.capacity() / 4 > out.z.len() {
+        out.z.shrink_to_fit();
+        out.x.shrink_to_fit();
+        out.phases.shrink_to_fit();
+        out.coeffs.shrink_to_fit();
+    }
+    Ok(out)
+}
+
+// This generates lookup tables of the form
+//      const LOOKUP: [[bool; 2] 4] = [[false, false], [true, false], [false, true], [true, true]];
+// when called `pauli_lookup!(LOOKUP, 2, [_, _])`.  The last argument is like a dummy version of
+// an individual lookup rule, which is consumed to make an inner "loop" with a declarative macro.
+macro_rules! pauli_lookup {
+    ($name:ident, $n:literal, [$head:expr$ (, $($tail:expr),*)?]) => {
+        static $name: [[bool; $n]; 1<<$n] = pauli_lookup!(@acc, [$($($tail),*)?], [[false], [true]]);
+    };
+    (@acc, [$head:expr $(, $($tail:expr),*)?], [$([$($bools:tt),*]),+]) => {
+        pauli_lookup!(@acc, [$($($tail),*)?], [$([$($bools),*, false]),+, $([$($bools),*, true]),+])
+    };
+    (@acc, [], $init:expr) => { $init };
+}
+pauli_lookup!(PAULI_LOOKUP_2, 2, [(), ()]);
+pauli_lookup!(PAULI_LOOKUP_4, 4, [(), (), (), ()]);
+pauli_lookup!(PAULI_LOOKUP_8, 8, [(), (), (), (), (), (), (), ()]);
+
+/// Push a complete Pauli chain into the output (`out`), if the corresponding entry is non-zero.
+/// `x` and `z` represent the symplectic X and Z bitvectors, packed into `usize`, where LSB n
+/// corresponds to qubit `n`.
+fn push_pauli_if_nonzero(
+    mut x: usize,
+    mut z: usize,
+    phase: u8,
+    value: Complex64,
+    out: &mut DecomposeOut,
+) {
+    if value.norm_sqr() <= out.tol {
+        return;
+    }
+
+    // This set of `extend` calls is effectively an 8-fold unrolling of the "natural" loop through
+    // each bit, where the initial `if` statements are handling the remainder (the up-to 7
+    // least-significant bits).  In practice, it's probably unlikely that people are decomposing
+    // 16q+ operators, since that's a pretty huge matrix already.
```

Comment:
Is it possible that an operator will have many qubits but with a few Pauli terms (sparse)?

Reply:
Yeah, very much so.  Everything here should handle those cases fine, though.